### PR TITLE
feat: support table ttl

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR allows user to create a table with time-to-live option. TTL defines how long data in some table may live. Rows with timestamp older than TTL will be marked deletable, and the actual deletion is defered to compaction.

During compaction, the picker finds out expired sst files before picking compaction candidates, so that expired ssts won't be involved.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#601

### CODE REVIEW

File: src/common/time/src/timestamp.rs [hunk 1]
```diff
@@ -122,15 +180,8 @@ impl Timestamp {
...
-        let (mut secs, mut nsecs) = self.split();
-
-        if nsecs < 0 {
-            secs -= 1;
-            nsecs += nano_factor;
-        }
-
-        Utc.timestamp_opt(secs, nsecs as u32)
+        let (sec, nsec) = self.split();
+        Utc.timestamp_opt(sec, nsec)
```
Comment (killme2008) [inline src/common/time/src/timestamp.rs:184]: Why changing the logic?

File: src/storage/src/compaction/picker.rs [hunk 2]
```diff
@@ -69,7 +94,18 @@ impl<S: LogStore> Picker for SimplePicker<S> {
...
         let levels = &req.levels();
+        let expired_ssts = self
+            .get_expired_ssts(levels, req.ttl)
+            .map_err(|e| {
+                error!(e;"Failed to get region expired SST files, region: {}, ttl: {:?}", req.region_id, req.ttl);
+                e
+            })
+            .unwrap_or(vec![]);
 
+        info!(
```
Comment (killme2008) [inline src/storage/src/compaction/picker.rs:105]: If the `expired_ssts` is empty, the log may be annoying. Only log it when it's not empty would be better.

File: src/common/time/src/timestamp.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/common/time/src/timestamp.rs:184]: > Why changing the logic?

To remain consistent with chrono's [`NaiveDateTime`](https://docs.rs/chrono/0.4.23/chrono/naive/struct.NaiveDateTime.html#method.from_timestamp). We are going to use chrono to do the timestamp calculation, including add/sub etc.

File: src/common/time/src/timestamp.rs [hunk 3]
```diff
@@ -17,20 +17,76 @@ use std::cmp::Ordering;
...
+};
 #[derive(Debug, Clone, Default, Copy, Serialize, Deserialize)]
 pub struct Timestamp {
     value: i64,
     unit: TimeUnit,
 }
 
 impl Timestamp {
+    /// Creates current timestamp in millisecond.
+    pub fn current_millis() -> error::Result<Self> {
```
Comment (evenyag) [inline src/common/time/src/timestamp.rs:40]: Just use `chrono::Utc::now().timestamp_millis()`?

File: src/common/time/src/timestamp.rs [hunk 4]
```diff
@@ -789,4 +840,41 @@ mod tests {
             Timestamp::new(1, TimeUnit::Second).convert_to_ceil(TimeUnit::Millisecond)
         );
     }
+
+    #[test]
+    fn test_split_overflow() {
+        let _ = Timestamp::new(i64::MAX, TimeUnit::Second).split();
```
Comment (evenyag) [inline src/common/time/src/timestamp.rs:846]: Could we assert the result here? Is it necessary to ignore the result explicitly?
```suggestion
        Timestamp::new(i64::MAX, TimeUnit::Second).split();
```

**Review by evenyag — COMMENTED** [review]
I realize that we should not return rows whose timestamps are expired during query. Maybe we could adjust the time_range in the query request by setting the start time to `min(expired_time, start)`

https://github.com/GreptimeTeam/greptimedb/blob/c6f2db8ae0265a7d31a3b49384bb7785ea570419/src/storage/src/chunk.rs#L154-L155

**Comment by waynexia** [comment]
>I realize that we should not return rows whose timestamps are expired during query.

I think it's not necessary. TTL is some kind of storage option, and can not take effort during query execution. If want to limit the selected ranges, specifying them in the query condition would be clearer.

**Comment by evenyag** [comment]
Looks like people can tolerate seeing expired data in query results. This saves lots of works 😄 

- https://ganeshvernekar.com/blog/prometheus-tsdb-compaction-and-retention/#time-based-retention
- https://clickhouse.com/docs/en/guides/developer/ttl/#triggering-ttl-events

**Comment by v0y4g3r** [comment]
> Looks like people can tolerate seeing expired data in query results. This saves lots of works 😄
> 
> TTL in some TSDB (prometheus) and analytical database (clickhouse) are also asynchronous.
> 
> * https://ganeshvernekar.com/blog/prometheus-tsdb-compaction-and-retention/#time-based-retention
> * https://clickhouse.com/docs/en/guides/developer/ttl/#triggering-ttl-events

As I descirbed in PR, we don't guarantee that expired data will be deleted immediately.

**Review by killme2008 — APPROVED** [review]
LGTM. Great work. Now we have the critical feature of TSDB.

File: src/common/time/src/timestamp.rs [hunk 5]
```diff
@@ -17,20 +17,66 @@ use std::cmp::Ordering;
...
 
 use chrono::offset::Local;
 use chrono::{DateTime, LocalResult, NaiveDateTime, TimeZone, Utc};
 use serde::{Deserialize, Serialize};
+use snafu::{OptionExt, ResultExt};
 
-use crate::error::{Error, ParseTimestampSnafu};
-
+use crate::error;
+use crate::error::{ArithmeticOverflowSnafu, Error, ParseTimestampSnafu, TimestampOverflowSnafu};
```
Comment (evenyag) [inline src/common/time/src/timestamp.rs:28]:
```suggestion
use crate::error::{ArithmeticOverflowSnafu, Error, ParseTimestampSnafu, TimestampOverflowSnafu};

```

File: src/storage/src/compaction/picker.rs [hunk 6]
```diff
@@ -69,6 +93,22 @@ impl<S: LogStore> Picker for SimplePicker<S> {
         req: &CompactionRequestImpl<S>,
     ) -> crate::error::Result<Option<CompactionTaskImpl<S>>> {
         let levels = &req.levels();
+        let expired_ssts = self
+            .get_expired_ssts(levels, req.ttl)
+            .map_err(|e| {
+                error!(e;"Failed to get region expired SST files, region: {}, ttl: {:?}", req.region_id, req.ttl);
+                e
+            })
+            .unwrap_or(vec![]);
```
Comment (evenyag) [inline src/storage/src/compaction/picker.rs:102]:
```suggestion
            .unwrap_or_default();
```

**Review by evenyag — COMMENTED** [review]
Maybe we could also check file expiration periodically.

**Comment by v0y4g3r** [comment]
> Maybe we could also check file expiration periodically.

That would be a good choice to spawn a task to check both expired and dangling SST files.

**Review by evenyag — APPROVED** [review]
LGTM

**Pull request merged**